### PR TITLE
Bundle extension fields through `custom`

### PR DESCRIPTION
**Pull request by technosophos**
This PR supersedes #28 

This adds a `custom` key as a designated place where extensions may be inserted into the `bundle.json` file.

See #28 for relevant discussion.

### CODE REVIEW

**Comment by technosophos** [comment]
This is mostly the same as #28, except it uses `custom` instead of `extensions`

File: 101-bundle-json.md [hunk 1]
```diff
@@ -500,4 +506,45 @@ The built-in actions (`install`, `upgrade`, `uninstall`) MUST NOT appear in the
 
 Implementations that do not support custom actions MUST NOT emit errors (either runtime or validation) if a bundle defines custom actions. That is, even if an implementation cannot execute custom actions, it MUST NOT fail to operate on bundles that declare custom actions.
 
+## Custom Extensions
+
+In many cases, the bundle descriptor is a sufficient artifact for delivering a CNAB bundle, since invocation images and other images may be retrieved from registries and repositories. Consequently, it is important to provide an extension mechanism. A _custom extension_ is a named collection of auxiliary data whose meaning is defined outside of this specification.
```
Comment (glyn) [inline 101-bundle-json.md]: Maybe "However" would be more appropriate than "Consequently" unless the previous sentence gave an example which _isn't_ covered by just the standard bundle descriptor data.

File: 101-bundle-json.md [hunk 2]
```diff
@@ -500,4 +506,45 @@ The built-in actions (`install`, `upgrade`, `uninstall`) MUST NOT appear in the
...
+
+Tools MAY define and declare additional fields inside of the `custom` section. Tools MUST NOT define additional fields anywhere else in the bundle descriptor. Implementations may produce an error or may ignore additional fields outside of the extension. However, implementations SHOULD preserve the data inside of the `custom` section even when that information is not understood by the implementation.
```
Comment (glyn) [inline 101-bundle-json.md]: Reword third sentence to: "Implementations may ignore additional fields outside of the extension or may produce an error when such fields are present."

Comment (silvin-lubecki) [inline 101-bundle-json.md]: nit: MAY in upper case

File: 101-bundle-json.md [hunk 3]
```diff
@@ -500,4 +506,45 @@ The built-in actions (`install`, `upgrade`, `uninstall`) MUST NOT appear in the
...
+    "extensions": {
+        "EXTENSION NAME": "ARBITRARY JSON DATA"
+    }
+}
+```
+
+The fields are defined as follows:
+
+- `extensions` defines the wrapper object for extensions
+  - EXTENSION NAME: a unique name for an extension. Names SHOULD follow the reverse DNS format `DOMAIN.ADDITIONAL.INFORMATION`, where DOMAIN conforms to the reverse DNS naming convention (`com.example` or `io.cnab`), and additional elements are separated by dots (`.`).
```
Comment (silvin-lubecki) [inline 101-bundle-json.md]: Should we factorize this part so we can reference it wherever we need namespaced names? (custom actions, labels, extensions...)

File: 101-bundle-json.md (hunk 1 above)
Comment (technosophos) [reply 101-bundle-json.md]: Fixed

File: 101-bundle-json.md (hunk 2 above)
Comment (technosophos) [reply 101-bundle-json.md]: Reworded to explain that this is an exclusive disjunction.

Comment (technosophos) [reply 101-bundle-json.md]: Fixed

File: 101-bundle-json.md (hunk 3 above)
Comment (technosophos) [reply 101-bundle-json.md]: Done.

**Pull request merged**